### PR TITLE
Add a charity contract demo

### PR DESCRIPTION
**Pull request by yucaowang**
## Description

This PR shows a contract demo of charity.
The contract could be controlled by a charity fund, and the owner is responsible to record all donates and costs, all people have the ability to review the detailed info of donates' and costs' records.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Brief of your solution

Provide `donate` and `cost` functions for the administrator of the fund to add records. 
Provide `queryDonor` function to check a donor's donation history.
Provide `queryDonates` and `queryCosts` function to check the detailed list of donates and costs.
Provide `statistics` function to query fund balance.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

### CODE REVIEW

File: core/contractsdk/cpp/example/charity_demo.cc [hunk 1]
```diff
@@ -0,0 +1,418 @@
+#include <iomanip>
+#include <sstream>
+#include "xchain/xchain.h"
+
+// 慈善捐款去向公式模板
```
Comment (qizheng09) [inline core/contractsdk/cpp/example/charity_demo.cc]: "公式" or "公示"

File: core/contractsdk/cpp/example/charity_demo.cc [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+// 参数由xchain::Contract中的context提供
+class Charity {
+public:
+    /*
+     * func: 初始化装备发行账户
```
Comment (qizheng09) [inline core/contractsdk/cpp/example/charity_demo.cc]: "装备"

Comment (yucaowang) [reply core/contractsdk/cpp/example/charity_demo.cc]: fixed

File: core/contractsdk/cpp/example/charity_demo.cc (hunk 1 above)
Comment (yucaowang) [reply core/contractsdk/cpp/example/charity_demo.cc]: fixed

File: core/contractsdk/cpp/example/charity_demo.cc [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+        const std::string& amountStr = ctx->arg("amount");
+        if (amountStr.empty()) {
+            ctx->error("missing 'amount'");
+            return;
+        }
+
+        uint64_t amount;
+        if (!safe_stoull(amountStr, &amount)) {
+            ctx->error(
```
Comment (ToWorld) [inline core/contractsdk/cpp/example/charity_demo.cc]: why two lines?

Comment (yucaowang) [reply core/contractsdk/cpp/example/charity_demo.cc]: to keep each line shorter than 80 characters.

**Pull request merged**